### PR TITLE
Fix movies with the same name matching wrong

### LINKED ISSUE #404: Movie sync title fallback matches different movies
I keep having issues syncing movies with identical names to trakt. If I see correctly it comes down to the title fallback coming in when they are actually not the same movie.

I have collected and seen the movie "[Blind (2014)](https://trakt.tv/movies/blind-2014)" and also collected "[Blind (2017)](https://trakt.tv/movies/blind-2017)". The latter will never get added on trakt because `compareMovies / findMediaObject` will end up going after the former which is already added on trakt.

This ends up being the matches:

(Kodi item to add)
``
DEBUG: [script.trakt] resources.lib.utilities: movie_col1 {'rating': 0, 'plays': 0, u'title': u'Blind', u'year': 2017, 'ids': {'imdb': u'tt1935089'}, 'watched': 0, 'watched_at': u'', u'userrating': 0, 'collected_at': u'2017-08-28 02:51:51+00:00', 'collected': 1, u'movieid': 15580, u'runtime': 5880}
``

(trakt item matched to)
``
DEBUG: [script.trakt] resources.lib.utilities: movie_col2 {'rating': 7, 'plays': 1, 'title': u'Blind', 'last_watched_at': '2014-07-05T16:42:12.000-00:00', 'year': 2014, 'ids': {'tmdb': '245844', 'imdb': 'tt2616810', 'slug': 'blind-2014', 'trakt': '149618'}, 'rated_at': '2014-07-20T14:39:17.000-00:00', 'progress': None, 'collected_at': '2014-09-06T23:34:07.000-00:00', 'collected': 1, 'watched': 1, 'in_watchlist': 0, 'paused_at': None}
``

(I have plenty of examples for this 😃... it can get as mainstream as "Wonder Woman".)

Imho the end result should not be a title match if the IDs IMDB is present and they are not matching. Would this make sense to add as a requirement at `findMediaObject` title match (or if anyone could add this as I'm a noob at python)?

### PR DESCRIPTION
**Pull request by razzeee**
@sp1ti

Closes #404

### CODE REVIEW

**Review by sp74fg2 — COMMENTED** [review]
Aye, have not yet started doing anything due to other activities. This certainly fixes title but different year but not different IDs from each other. Ex. tt0402910 tt0405977. 
What I would have done is to check if IMDB id is present and if yes match them against each other. But cases of title+year being identical are about as few as duplicate entries on IMDB (which do exist also) 👍  Thanks

**Comment by razzeee** [comment]
The thing is your problem is collecting right?

We end up with two lists:
a) List of local shows with both versions of blind
b) List of remote/trakt shows only with the newer blind (as it's not added to the collection yet)

We match these. If list b has both blind movies it will match both fine by imdb id. If like in your case it's missing, it's falling back to title search, matches and will never be added as it thinks you already have it.

**Comment by sp74fg2** [comment]
Yes my issue is with collecting. I followed the matching rules to that point yes but it would still not push one or the other if they have the same title and year despite having a different IMDBid (and thus being different movies). 

A few examples:
"Chaos" (2005) tt0402910 tt0405977
"Brothers" (2015) (on TMDB) tt2835658 tt5803522
"Escape Room" (2017) tt5748392 tt5159414

**Review by sp74fg2 — APPROVED** [review]

**Comment by razzeee** [comment]
Yes, same year is still a problem. https://github.com/Razzeee/script.trakt/blob/d9cb536b9d6677e813fbc1482c440409b5faee5e/resources/lib/utilities.py#L104

Not 100% sure what to to about that.

**Comment by sp74fg2** [comment]
Can the __findInList have filters for "not equal"? If so check that they're not matching the mediaObjectToMatch['ids'].

**Comment by razzeee** [comment]
I've added some more tests and removed the title+year matching. It seems like I inherited it from the old implementation. So getting rid of it is a risk I'm willing to take.

Some testing would be appreciated.

**Comment by sp74fg2** [comment]
Ok, removing it sort of makes sense I guess 😄  (since people rely on scrappers having those IDs and it avoids wrong data). The only issue I see is that trakt from time to time missed some TMDB changes and not always has an IMDBid (I don't think the TMDBid is ever queried from kodi). 
I will test this tomorrow with my data/profile. Thanks 💘

**Comment by razzeee** [comment]
kodi has some big changes with v17 regarding ids. But most scrappers do not support those yet. But whenever that feature lands we can start sending those ids immediately.

Edit: My small testsetup seems fine. But is probably missing exotic movies/shows.

**Comment by sp74fg2** [comment]
Wouldn't this also be missing from the JSON-RPC API? Cannot really see anything else in forms of IDs.

**Comment by razzeee** [comment]
Not sure what your saying.

**Comment by sp74fg2** [comment]
You are requesting the library data via the JSON-RPC API. But there isn't any field for "IDs" except for `imdbnumber` and `uniqueid`? http://kodi.wiki/view/JSON-RPC_API/v8#VideoLibrary.GetMovies

**Comment by razzeee** [comment]
It's in `uniqueid`. 
See: http://kodi.wiki/view/JSON-RPC_API/v8#Media.UniqueID
And for context https://github.com/xbmc/xbmc/pull/10000

**Review by sp74fg2 — APPROVED** [review]
Tested OK (movies).

**Comment by sp74fg2** [comment]
Sorry for the later reply than intended: works well - thank you! It added a bit more than 200 titles. including the example of "Blind" and things like "The Girl on the Train". 
The downside is that now TMDB only movies are added to the list of movies it supposedly adds (went from 153 each cycle to now 724). I'm not representative for the average user but something to keep in mind if this gets implemented.

Also thanks for the uniqueID info; this will hopefully be added to the scrappers around.

**Comment by razzeee** [comment]
Can you explain the TMDB part? 153 to 724 for eg, don't get it right now.

**Comment by sp74fg2** [comment]
Each sync cycle I have movies repeatedly shown to me as "will be added" but never will be.

ex. [Movies Sync] Compared movies, found 158 to add.
[Movies Sync] 158 movie(s) will be added to Trakt.tv collection.

The main reasons for that is: 
- No local meta data (neither IMDB or TMDB)
- trakt is missing the movie either because it was not added to TMDB yet OR because the trakt data is outdated and never grabbed an updated IMDB from TMDB
- trakt blocks the import (adult, edgy movies, mini-series as a movie etc.)

Because it now won't fall back on year or title matching it presents me with additional movies it tries to add but cannot actually do for the reasons above (scrapper lacking TMDBid implementation). I have quite a few with only TMDBids which used to match and now don't. That is why my count jumped by a good 500 with the change implemented.

Examples:
https://trakt.tv/movies/death-in-june-behind-the-mask-2005
https://trakt.tv/movies/fuuin-eizou-8-jishou-joou-2012
https://trakt.tv/movies/mushishi-the-next-chapter-drops-of-bells-2015

It's not an issue for me as this will resolve once scrappers adapt. I prefer more exact matches any day (I just went to see "Animals" and "The Square" this week which both added wrong plays back since I have other movies called this in my collection - this won't happen anymore).

**Comment by razzeee** [comment]
@rectifyer an the api answer with a reason why something didn't get added?

**Comment by sp74fg2** [comment]
There is a response that could be processed (blocked is not actually a status if that is what you are referring to. it just means that the item is either hidden on trakt or won't let itself be imported from TMDB.):
http://docs.trakt.apiary.io/#reference/sync/add-to-collection/add-items-to-collection


In my ideal world the transparency level with the addon would be:
1. A log record for every movie that will be skipped due to missing meta data (IDs) so that I know which ones need to be added at TMDB (which at the moment will be submitted regardless).
2. The current submission log as it is (ideally reinclude the IDs as in the past (I'll try a pull request for this sometime in the future to make it more like the TV log)). This allows me to check what is not picking up from trakt (be it stale IDs or blocked).
3. Logged response body from trakt that is also handled with the GUI so that it does not need to lie to me (X added / Y skipped / Z not found).

But as I said; I can work around the current limitations as I've been doing for a long time. I'm happy with just having "IDs" respected so everything actually syncs that can and it won't be hooking up the wrong movies together.

**Comment by rectifyer** [comment]
@Razzeee It will return in the `not_found` object if it can't match on any of the IDs or title. Check out http://docs.trakt.apiary.io/#reference/sync which explains the order it uses in the **Media objects for syncing** section. Usually it can be fixed on TMDB (and thus next day on Trakt after it syncs), but if needed a Trakt support ticket can be opened for that specific movie and we can look into getting it fixed with the missing IDs.

**Comment by razzeee** [comment]
Sounds like a lot of manual work. :(
Does it also return not found for adault movies that can't be added anyway?

**Comment by razzeee** [comment]
@rectifyer

File: resources/lib/utilities.py [hunk 1]
```diff
@@ -100,15 +100,6 @@ def findMediaObject(mediaObjectToMatch, listToSearch):
     if result is None and 'ids' in mediaObjectToMatch and 'tvdb' in mediaObjectToMatch['ids'] and mediaObjectToMatch['ids']['tvdb']:
         result = __findInList(
             listToSearch, tvdb=mediaObjectToMatch['ids']['tvdb'])
-    # match by title and year it will result in movies with the same title and
```
Comment (razzeee) [inline resources/lib/utilities.py:103]: Better idea, bind this to the setting for name matching fallback?

**Comment by rectifyer** [comment]
@Razzeee yes, it would return not found for adult movies as well. Basically anything we can't find a match for in Trakt (badly formed data, invalid id, adult movie, etc).

**Comment by razzeee** [comment]
Feels like I read this different a month ago (or in a hurry...). Thanks for clearing it up.

File: resources/lib/utilities.py (hunk 1 above)
Comment (sp74fg2) [reply resources/lib/utilities.py:103]: Forgot about this setting but yes, that would make the most sense!

Comment (razzeee) [reply resources/lib/utilities.py:103]: Same, totally forgot about it. Saw some discussion on another PR yesterday, which got me thinking we might break some usecase when throwing out match by name completely.

**Comment by razzeee** [comment]
@sp1ti btw multiple id scrapers have landed, at least for krypton and up. I think only tmdb movies for now, but I'll keep pushing for tvdb.

File: resources/lib/utilities.py (hunk 1 above)
Comment (razzeee) [reply resources/lib/utilities.py:103]: Done btw

**Pull request merged**